### ISSUE TITLE
Warn when including CachedRender in a routing target or root component

### ISSUE THREAD

**Issue by jgaskins** [clearwater-rb/clearwater#43]
The `CachedRender` mixin can cause problems when included by a component that is reused between renders. Specifically, their cache will never be invalidated because `should_render?` will never be invoked. This is an optimization within `virtual-dom` that we cannot control at our level.

Since the application root component and all routing targets are examples of these, they should not include this mixin. We can determine this at app load time and provide a good warning message.

**Comment by jiayp** [comment]
@jgaskins Can we cache the render result in Component.sanitize_content like this
```
              if !content.cached_node || content.$should_render?(content.cached_state)
                content.cached_node = self.$sanitize_content(content.$render())
              end
              content.cached_state = content.$save_state()
              return content.cached_node
```
This will call `should_render?` no matter if people cached the Component instance,and render correctly.
Current CachedRender request you must not "cached" the Component instance.It's wired.

**Comment by jgaskins** [comment]
@jiayp I thought about this a bit while I was out walking the dog. I'll have to test it to be sure, but I think the bigger problem is that if we don't use the `virtual-dom` hook, we won't be able to get the previous version of the component to compare it to.

I agree that it's weird. It caused a few problems for me right after I made the switch to `virtual-dom` thunks until I realized you couldn't cache the component and again when I tried to upgrade a Clearwater app that I had from back when render caching *had* to be done by caching the component instance on the parent component (all the way up to the root or nearest routing target).

**Comment by jiayp** [comment]
In my use case,I cached the render node to my models.

I have a model list and many models,I don't want maintenance a respond cache list in my view component.So I cache the render node to model.

If we just pass the render node to component,and let the derived component handle the cache strategy?

**Comment by jgaskins** [comment]
I'm not sure putting view concerns on the domain models is the right way to go. It might not be a bad idea for your specific app, but in general I like to keep view-related things inside the component.

You can still pass a rendered node to the component if it works for you, but being able to invalidate the cache based on whether the inputs to the component (the model, for example, or even certain attributes on the model) have changed is pretty powerful.

**Comment by jgaskins** [comment]
Fixed in f9183adfe24c3f1d87f3c5ce0008afc3e7de3a7e
